packages/flutter_tools/lib/src/macos/xcdevice.dart

```diff
@@ -198,51 +212,101 @@ class XCDevice {
-        .transform<String>(utf8.decoder)
-        .transform<String>(const LineSplitter())
-        .listen((String line) {
-        final XCDeviceEventNotification? event = _processXCDeviceStdOut(
-          line,
-          XCDeviceEventInterface.usb,
-        );
-        if (event != null) {
-          _deviceIdentifierByEvent?.add(<XCDeviceEvent, String>{
-            event.eventType: event.deviceIdentifier,
-          });
-        }
+      _wifiDeviceObserveProcess = await _processUtils.start(
+        <String>[
+          'script',
+          '-t',
+          '0',
+          '/dev/null',
+          ..._xcode.xcrunCommand(),
+          'xcdevice',
+          'observe',
+          '--${XCDeviceEventInterface.wifi.name}',
+        ],
+      );
+
+      final StreamSubscription<String> usbStdoutSubscription = _processObserveStdOut(
+        _usbDeviceObserveProcess!,
+        XCDeviceEventInterface.usb,
+      );
+      final StreamSubscription<String> wifiStdoutSubscription = _processObserveStdOut(
+        _wifiDeviceObserveProcess!,
+        XCDeviceEventInterface.wifi,
+      );
+
+      final StreamSubscription<String> usbStderrSubscription = _processXCDeviceStdErr(
+        _usbDeviceObserveProcess!,
+        XCDeviceCommand.observe,
+      );
+      final StreamSubscription<String> wifiStderrSubscription = _processXCDeviceStdErr(
+        _wifiDeviceObserveProcess!,
+        XCDeviceCommand.observe,
+      );
+
+      final Future<void> usbProcessExited = _usbDeviceObserveProcess!.exitCode.then((int status) {
+        _logger.printTrace('xcdevice observe --usb exited with code $exitCode');
+        // Kill other process in case only one was killed.
+        _wifiDeviceObserveProcess?.kill();
+        unawaited(usbStdoutSubscription.cancel());
+        unawaited(usbStderrSubscription.cancel());
       });
-      final StreamSubscription<String> stderrSubscription = _deviceObservationProcess!.stderr
-        .transform<String>(utf8.decoder)
-        .transform<String>(const LineSplitter())
-        .listen((String line) {
-        _logger.printTrace('xcdevice observe error: $line');
+
+      final Future<void> wifiProcessExited = _wifiDeviceObserveProcess!.exitCode.then((int status) {
+        _logger.printTrace('xcdevice observe --wifi exited with code $exitCode');
+        // Kill other process in case only one was killed.
+        _usbDeviceObserveProcess?.kill();
+        unawaited(wifiStdoutSubscription.cancel());
+        unawaited(wifiStderrSubscription.cancel());
       });
-      unawaited(_deviceObservationProcess?.exitCode.then((int status) {
-        _logger.printTrace('xcdevice exited with code $exitCode');
-        unawaited(stdoutSubscription.cancel());
-        unawaited(stderrSubscription.cancel());
-      }).whenComplete(() async {
-        if (_deviceIdentifierByEvent?.hasListener ?? false) {
+
+      unawaited(Future.wait(<Future<void>>[
```

Comment:
What if only one dies?

Reply:
When one dies, it also kills the other
https://github.com/flutter/flutter/blob/55f5d058e6c62ea136ced09b74c5e07585b5fad0/packages/flutter_tools/lib/src/macos/xcdevice.dart#L250-L253